lib/collection/src/shards/shard_holder/mod.rs

```diff
@@ -1119,13 +1119,37 @@ impl ShardHolder {
         Ok(())
     }
 
-    /// Queries and accumulates the statistics for local data, uncached.
-    pub async fn calculate_local_shards_stats(&self) -> LocalDataStats {
-        let mut stats = LocalDataStats::default();
+    /// Estimates the collections size based on local shard data. Returns `None` if no shard for the collection was found locally.
+    pub async fn estimate_collection_size_stats(&self) -> Option<CollectionSizeStats> {
+        if self.is_distributed().await {
+            // In distributed, we estimate the whole collection size by using a single local shard and multiply by amount of shards in the collection.
+            for shard in self.shards.iter() {
+                if let Some(shard_stats) = shard.1.calculate_local_shard_stats().await {
+                    // Project the single shards size to the full collection.
+                    let collection_estimate = shard_stats.multiplied_with(self.shards.len());
```

Comment:
@timvisee would this approach break while running resharding?

I am afraid the estimate would be much larger/smaller while we transfer data to new shards.

Reply:
> would this approach break while running resharding?

Yes.

It'll not be that easy to do while resharding is going on. Depending on the current state of resharding, the calculation will be different. If there are multiple resharding operations going on at the same time, this will become more complex.

Would it be possible to sum each shard separately for now, and improve on this in a separate PR?

Reply:
@timvisee The multiplication approach relies on the fact that all shards have the assumption that all shards have more or less the same size. Are there maybe some resharding state we could leverage to refine our estimation? 

Without the multiplication trick, we have a hole in the estimation because we will only account for what is present locally.

@JojiiOfficial I am now a bit confused with the naming `estimate_collection_size_stats`.
Are we just interested in what is "local" on the node or are trying to have actual estimate of the full collection size?

For the full collection size case, we could leverage our `ReplicaSet` abstraction and multiply the size of the local shard by the number of remote replicas.

I understand that If we have zero local shard, we assume the internal API on the remote shard will be responsible to computing a the collection size.

Reply:
In a call we agreed it might be best to estimate like this: `shard_size * shard IDs * replication factor`

We use the replication factor and now the real number of replicas because the 'cost for extra replicas during transfer' is on us. The replication factor better represents what data we care about.

Reply:
I would vote for not including replication factor at all

Reply:
@generall Could you please elaborate on why you'd rather not account for replication factor?

My guess is that the replication factor an implementation details in the serverless setup.
But what about the non serverless use case then?

Reply:
we might want to temporary create additional replicas for the purpose of data migration. In serverless, user have no control over replication factor. I would consider this size limitation as a limitation for the size of unique data, which doesn't include duplication from segments or shards

Reply:
> I would consider this size limitation as a limitation for the size of unique data, which doesn't include duplication from segments or shards

:+1:

Reply:
The approach of this PR doesn't include replicas. I've updated the strict mode options documentation to make this more clear.
We'll adjust this estimation in future to eg. make it compatible with resharding.